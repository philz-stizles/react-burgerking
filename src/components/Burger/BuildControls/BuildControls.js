import React from 'react'
import BuildControl from './BuildControl/BuildControl'
import './BuildControls.css'

const controls = [
    { label: 'Salad', type: 'salad'},
    { label: 'Bacon', type: 'bacon'},
    { label: 'Cheese', type: 'cheese'},
    { label: 'Meat', type: 'meat'}
]

export const BuildControls = ({ onAddIngredient, onRemoveIngredient, disabledStates, totalPrice, purchasable, onShowModal}) => {
    return (
        <div className="BuildControls">
            <p>Current Price: <strong>{totalPrice.toFixed(2)}</strong></p>
            {
                controls.map((control,  i) => {
                    return <BuildControl key={i} 
                        {...control} 
                        onAddIngredient={() => onAddIngredient(control.type)} 
                        onRemoveIngredient={() => onRemoveIngredient(control.type)}
                        isDisabled={disabledStates[control.type]}
                    />
                })
            }
            <button onClick={onShowModal} className="OrderButton" disabled={!purchasable}>ORDER NOW</button>
        </div>
    )
}
