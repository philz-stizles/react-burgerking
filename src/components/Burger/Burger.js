import React from 'react'
// import Radium from 'radium'
import './Burger.css'
import BurgerIngredient from './BurgerIngredient/BurgerIngredient'

const Burger = (props) => {
    const { ingredients } = props
    let transformedIngredients = Object.keys(ingredients).map(ingredient => {
        return [...Array(ingredients[ingredient])].map((_, i) => {
            return <BurgerIngredient key={`${ingredient}${i}`} type={ingredient} />
        })
    })

    const totalIngredients = transformedIngredients.reduce((accumulator, item) => {
        return accumulator + item.length
    }, 0)

    if(totalIngredients <= 0) {
        transformedIngredients = <p>Please start adding incredients</p>
    }
    
    return (
        <div className="Burger">
            <BurgerIngredient type="bread-top" />
            { transformedIngredients }
            <BurgerIngredient type="bread-bottom" />
        </div>
    )
}

export default Burger