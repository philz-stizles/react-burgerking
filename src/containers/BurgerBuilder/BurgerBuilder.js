import React, { Component } from 'react'
import { BuildControls } from '../../components/Burger/BuildControls/BuildControls'
import Burger from '../../components/Burger/Burger'
import Auxi from '../../hoc/Auxilliary/Auxilliary'
import Modal from '../../components/UI/Modal/Modal'
import OrderSummary from '../../components/Burger/OrderSummary/OrderSummary'
import Spinner from '../../components/UI/Spinner/Spinner'
import axios from '../../api/axios-orders'
import withErrorHandler from '../../hoc/withErrorHandler/withErrorHandler'

const INGREDIENT_PRICES = {
    salad: 0.5,
    bacon: 0.4,
    cheese: 1.3,
    meat: 0.7
}

class BurgerBuilder extends Component {
    state = {
        ingredients: null,
        totalPrice: 0,
        purchasable: false,
        showModal: false,
        loading: false,
        error: false
    }

    componentDidMount() {
        axios.get('/ingredients.json')
            .then(response => {
                console.log(response)
                this.setState({ingredients: response.data})
            })
            .catch(error => {
                this.setState({error: true})
            })
    }

    addIngredientHandler = (type) => {
        const { ingredients, totalPrice } = this.state
        const updatedIngredients = {
            ...ingredients,
            [type]: ingredients[type] + 1
        }
        const count = this.updatePurchasable(updatedIngredients)
        this.setState({
            ingredients: updatedIngredients,
            totalPrice: totalPrice + INGREDIENT_PRICES[type],
            purchasable: (count <= 0) ? false : true
        })
    }

    removeIngredientHandler = (type) => {
        const { ingredients, totalPrice } = this.state
        const updatedIngredients = {
            ...ingredients,
            [type]: (ingredients[type] <= 0) ? 0 : ingredients[type] - 1
        }
        const count = this.updatePurchasable(updatedIngredients)
        this.setState({
            ingredients: updatedIngredients,
            totalPrice: totalPrice - INGREDIENT_PRICES[type],
            purchasable: (count <= 0) ? false : true
        })
    }

    updatePurchasable = (ingredients) => {
        return Object.keys(ingredients)
            .map(ingredientKey => ingredients[ingredientKey])
            .reduce((accumulator, ingredientValue) => accumulator + ingredientValue, 0)
    }

    toggleModal = () => {
        this.setState({
            showModal: !this.state.showModal
        })
    }

    closeModal = () => {
        this.setState({
            showModal: false
        })
    }

    orderCancel = () => {
        console.log('cancel')
        this.setState({showModal: false})
    }

    goToCheckout = () => {
        const { ingredients } = this.state
        const queryParams = Object.keys(ingredients)
            .map(ingredient => {
                return `${encodeURIComponent(ingredient)}=${encodeURIComponent(ingredients[ingredient])}`
            })
            .join('&')
        this.setState({showModal: false})
        this.props.history.push({
            pathname: '/checkout',
            search: queryParams
        })

        // this.setState({loading: true})

        // axios.post('/orders.json', ingredients)
        //     .then(response => {
        //         console.log(response)
        //         this.setState({loading: false, showModal: false})
        //     })
        //     .catch(error => {
        //         console.log(error)
        //         this.setState({loading: false, showModal: false})
        //     })
    }

    render() {
        const { ingredients, totalPrice, purchasable, showModal, loading, error } = this.state
        const disabledStates = {}
        for(let key in ingredients) {
            disabledStates[key] = (ingredients[key] <= 0 ) ? true : false
        }

        let burgerComponentsOrSpinner = error ? <p>Ingredients can't be loaded</p> : <Spinner />
        let modalComponent = null

        if(ingredients) {
            burgerComponentsOrSpinner = (
                <Auxi>
                    <Burger ingredients={ingredients}/>
                    <BuildControls 
                        purchasable={purchasable}
                        totalPrice={totalPrice}
                        onAddIngredient={this.addIngredientHandler} 
                        onRemoveIngredient={this.removeIngredientHandler}
                        disabledStates={disabledStates}
                        onShowModal={this.toggleModal}
                    />
                </Auxi>
            )

            modalComponent = <OrderSummary 
                ingredients={ingredients} 
                orderCancel={this.orderCancel} 
                orderContinue={this.goToCheckout}
                totalPrice={totalPrice}
            />

            if(loading) {
                modalComponent = <Spinner />
            }
        }

        return ( 
            <Auxi>
                <Modal show={showModal} onModalClose={this.closeModal}>
                    {modalComponent}
                </Modal>
                {
                    burgerComponentsOrSpinner
                }
            </Auxi>
        )
    }
}

export default withErrorHandler(BurgerBuilder, axios)