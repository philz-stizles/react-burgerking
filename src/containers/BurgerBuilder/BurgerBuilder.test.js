import React from 'react'
import { shallow } from 'enzyme'
import BurgerBuilder from './BurgerBuilder'
import { BuildControls } from '../../components/Burger/BuildControls/BuildControls'
import Burger from '../../components/Burger/Burger'

describe('<BurgerBuilders />', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallow(<BurgerBuilder />)
    })

    it('should render <BurgerBuilders />', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    it('should render one <BuildControls />', () => {
        expect(wrapper.find(BuildControls)).toHaveLength(1)
    })

    it('should render one <Burger />', () => {
        expect(wrapper.find(Burger)).toHaveLength(1)
    })

    it('has initial state', () => {
        expect(wrapper.state()).toEqual({
            ingredients: {
                salad: 0,
                bacon: 0,
                cheese: 0,
                meat: 0
            },
            totalPrice: 0,
            purchasable: false,
            showModal: false,
            loading: false
        })
    })

    it('initial state', () => {
        expect(wrapper.state()).toEqual({
        })
    })
})