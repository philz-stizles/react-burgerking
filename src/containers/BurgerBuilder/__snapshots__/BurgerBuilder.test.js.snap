// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BurgerBuilders /> should render <BurgerBuilders /> 1`] = `ShallowWrapper {}`;
